velox/connectors/hive/HiveWriteProtocol.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/HiveWriteProtocol.h"
+#include "velox/connectors/hive/HiveConnector.h"
+
+#include <boost/lexical_cast.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace facebook::velox::connector::hive {
+
+namespace {
+
+std::vector<std::shared_ptr<const HiveColumnHandle>> getPartitionColumns(
+    const std::shared_ptr<const HiveInsertTableHandle>& tableWriteHandle) {
+  std::vector<std::shared_ptr<const HiveColumnHandle>> partitionColumns;
+  for (const auto& column : tableWriteHandle->inputColumns()) {
+    const auto hiveColumn =
+        std::dynamic_pointer_cast<const HiveColumnHandle>(column);
+    if (hiveColumn->columnType() ==
+        HiveColumnHandle::ColumnType::kPartitionKey) {
+      partitionColumns.emplace_back(hiveColumn);
+    }
+  }
+  return partitionColumns;
+}
+
+std::string makeUuid() {
+  return boost::lexical_cast<std::string>(boost::uuids::random_generator()());
+}
+
+} // namespace
+
+std::shared_ptr<const WriterParameters>
+HiveNoCommitWriteProtocol::getWriterParameters(
+    const std::shared_ptr<const ConnectorInsertTableHandle>& tableWriteHandle,
+    const ConnectorQueryCtx* FOLLY_NONNULL connectorQueryCtx) const {
+  auto hiveTableWriteHandle =
+      std::dynamic_pointer_cast<const HiveInsertTableHandle>(tableWriteHandle);
+  if (!hiveTableWriteHandle) {
+    VELOX_UNSUPPORTED("This write protocol cannot used for non-Hive connector");
+  }
+
+  if (!getPartitionColumns(hiveTableWriteHandle).empty()) {
+    VELOX_NYI(
+        "Getting write parameters for partitioned Hive tables is not implemented yet.");
+  }
+
+  VELOX_USER_CHECK(
+      hiveTableWriteHandle->isCreateTable() ||
+          !connectorQueryCtx->config()->get("hive.immutable-partitions", false),
+      "Unpartitioned Hive tables are immutable");
+
+  auto targetFileName = fmt::format(
+      "{}_{}_{}",
+      connectorQueryCtx->taskId(),
+      connectorQueryCtx->driverId(),
+      0);
+  auto writeFileName = targetFileName;
+
+  return std::make_shared<HiveWriterParameters>(
```

Comment:
Would it make sense to modify HiveWriterParameters to make writeXxx properties std::optional, so that if not specified they assumed to be the same as targetXxx?

Reply:
Yes